### PR TITLE
Support watching a single node

### PR DESCRIPTION
**Pull request by caseydavenport**
## Description

Fixes https://github.com/projectcalico/calico/issues/1739

## Todos
- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
None required
```

### CODE REVIEW

**Comment by shafay2000** [comment]
Thanks Casey!

File: lib/backend/k8s/resources/node.go [hunk 1]
```diff
@@ -160,11 +160,16 @@ func (c *nodeClient) EnsureInitialized() error {
...
 func (c *nodeClient) Watch(ctx context.Context, list model.ListInterface, revision string) (api.WatchInterface, error) {
-	if len(list.(model.ResourceListOptions).Name) != 0 {
-		return nil, fmt.Errorf("cannot watch specific resource instance: %s", list.(model.ResourceListOptions).Name)
+	// Build watch options to pass to k8s.
+	opts := metav1.ListOptions{ResourceVersion: revision, Watch: true}
+	name := list.(model.ResourceListOptions).Name
+	if len(name) != 0 {
+		// We've been asked to watch a specific node resource.
+		log.WithField("name", name).Debug("Watching a single node")
+		opts.FieldSelector = fields.OneTermEqualSelector("metadata.name", name).String()
```
Comment (robbrockbank) [inline lib/backend/k8s/resources/node.go]: hopefully a quick question.  Does the API server handle this field selector explicitly by constructing the correct watch path for etcdv3, or does this actually end up watching everything, but filter out all other nodes at the API server?

Comment (caseydavenport) [reply lib/backend/k8s/resources/node.go]: I don't know the apiserver code super well, but my quick read of it led me to believe that it watches just this one specific object. It seems aware that we've requested a watch on the name of the resource, and has special case code for that.

Even if it internally watched all keys and filtered, that would still be nicer than re-querying every 5 seconds.

Comment (caseydavenport) [reply lib/backend/k8s/resources/node.go]: https://github.com/kubernetes/apiserver/blob/release-1.9/pkg/registry/generic/registry/store.go#L1152-L1177

File: lib/backend/k8s/resources/node.go [hunk 2]
```diff
@@ -160,11 +160,16 @@ func (c *nodeClient) EnsureInitialized() error {
 }
 
 func (c *nodeClient) Watch(ctx context.Context, list model.ListInterface, revision string) (api.WatchInterface, error) {
-	if len(list.(model.ResourceListOptions).Name) != 0 {
-		return nil, fmt.Errorf("cannot watch specific resource instance: %s", list.(model.ResourceListOptions).Name)
+	// Build watch options to pass to k8s.
+	opts := metav1.ListOptions{ResourceVersion: revision, Watch: true}
+	name := list.(model.ResourceListOptions).Name
```
Comment (tmjd) [inline lib/backend/k8s/resources/node.go]: I see that we didn't check before but should we check that this conversion works?

File: lib/backend/k8s/k8s_test.go [hunk 3]
```diff
@@ -1769,3 +1769,65 @@ var _ = Describe("Test Syncer API for Kubernetes backend", func() {
...
+
+	It("Should support watching Nodes", func() {
+		By("Watching a single node", func() {
+			name := "127.0.0.1" // Node created by test/mock-node.yaml
+			watch, err := c.Watch(ctx, model.ResourceListOptions{Name: name, Kind: apiv3.KindNode}, "")
+			Expect(err).NotTo(HaveOccurred())
+
+			// We should get at least one event from the watch.
+			var receivedEvent bool
+			for i := 0; i < 10; i++ {
```
Comment (tmjd) [inline lib/backend/k8s/k8s_test.go]: Should we make a change to the resource and see that we get that in the test, to make sure that the watch is setup correctly?

File: lib/backend/k8s/resources/node.go (hunk 2 above)
Comment (caseydavenport) [reply lib/backend/k8s/resources/node.go]: Yes, I suppose that is safest.

File: lib/backend/k8s/k8s_test.go (hunk 3 above)
Comment (caseydavenport) [reply lib/backend/k8s/k8s_test.go]: I think this is enough to verify the watch is correct - we verify that we get at least one non-error event.

**Comment by tmjd** [comment]
Closing and reopening to see if that makes license/cia happy.

**Pull request merged**